# Ensemble Learning

## Initial Imports

In [98]:
import warnings
warnings.filterwarnings('ignore')

In [99]:
import numpy as np
import pandas as pd
from pathlib import Path
from collections import Counter

In [100]:
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced

## Read the CSV and Perform Basic Data Cleaning

In [101]:
# Load the data
file_path = Path('./Resources/LoanStats_2019Q1.csv')
df = pd.read_csv(file_path)

# Preview the data
df.head()

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0    10500.0    0.1719       375.35           RENT     66000.0   
1    25000.0    0.2000       929.09       MORTGAGE    105000.0   
2    20000.0    0.2000       529.88       MORTGAGE     56000.0   
3    10000.0    0.1640       353.55           RENT     92000.0   
4    22000.0    0.1474       520.39       MORTGAGE     52000.0   

  verification_status   issue_d loan_status pymnt_plan    dti  ...  \
0     Source Verified  Mar-2019    low_risk          n  27.24  ...   
1            Verified  Mar-2019    low_risk          n  20.23  ...   
2            Verified  Mar-2019    low_risk          n  24.26  ...   
3            Verified  Mar-2019    low_risk          n  31.44  ...   
4        Not Verified  Mar-2019    low_risk          n  18.76  ...   

   pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0            85.7             100.0                   0.0        0.0   
1            91.2              50.0                   1.0        0.0   
2            66.7              50.0                   0.0        0.0   
3           100.0              50.0                   1.0        0.0   
4           100.0               0.0                   0.0        0.0   

   tot_hi_cred_lim  total_bal_ex_mort total_bc_limit  \
0          65687.0            38199.0         2000.0   
1         271427.0            60641.0        41200.0   
2          60644.0            45684.0         7500.0   
3          99506.0            68784.0        19700.0   
4         219750.0            25919.0        27600.0   

   total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                     61987.0              N                     N  
1                     49197.0              N                     N  
2                     43144.0              N                     N  
3                     76506.0              N                     N  
4                     20000.0              N                     N  

[5 rows x 86 columns]

## Split the Data into Training and Testing

In [102]:
# Create our features
X = df.drop(columns='loan_status')

# Create our target
y = df['loan_status']

In [103]:
X.describe()

loan_amnt      int_rate   installment    annual_inc           dti  \
count  68817.000000  68817.000000  68817.000000  6.881700e+04  68817.000000   
mean   16677.594562      0.127718    480.652863  8.821371e+04     21.778153   
std    10277.348590      0.048130    288.062432  1.155800e+05     20.199244   
min     1000.000000      0.060000     30.890000  4.000000e+01      0.000000   
25%     9000.000000      0.088100    265.730000  5.000000e+04     13.890000   
50%    15000.000000      0.118000    404.560000  7.300000e+04     19.760000   
75%    24000.000000      0.155700    648.100000  1.040000e+05     26.660000   
max    40000.000000      0.308400   1676.230000  8.797500e+06    999.000000   

        delinq_2yrs  inq_last_6mths      open_acc       pub_rec  \
count  68817.000000    68817.000000  68817.000000  68817.000000   
mean       0.217766        0.497697     12.587340      0.126030   
std        0.718367        0.758122      6.022869      0.336797   
min        0.000000        0.000000      2.000000      0.000000   
25%        0.000000        0.000000      8.000000      0.000000   
50%        0.000000        0.000000     11.000000      0.000000   
75%        0.000000        1.000000     16.000000      0.000000   
max       18.000000        5.000000     72.000000      4.000000   

           revol_bal  ...  num_tl_90g_dpd_24m  num_tl_op_past_12m  \
count   68817.000000  ...        68817.000000        68817.000000   
mean    17604.142828  ...            0.052138            2.219423   
std     21835.880400  ...            0.390633            1.897432   
min         0.000000  ...            0.000000            0.000000   
25%      6293.000000  ...            0.000000            1.000000   
50%     12068.000000  ...            0.000000            2.000000   
75%     21735.000000  ...            0.000000            3.000000   
max    587191.000000  ...           18.000000           19.000000   

       pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
count    68817.000000      68817.000000          68817.000000    68817.0   
mean        95.057627         30.626217              0.125972        0.0   
std          8.326426         33.631463              0.336732        0.0   
min         20.000000          0.000000              0.000000        0.0   
25%         93.000000          0.000000              0.000000        0.0   
50%        100.000000         20.000000              0.000000        0.0   
75%        100.000000         50.000000              0.000000        0.0   
max        100.000000        100.000000              4.000000        0.0   

       tot_hi_cred_lim  total_bal_ex_mort  total_bc_limit  \
count     6.881700e+04       6.881700e+04    68817.000000   
mean      2.100332e+05       6.133843e+04    29734.128558   
std       1.928088e+05       5.738798e+04    26795.394232   
min       3.600000e+03       2.350000e+02      100.000000   
25%       6.697700e+04       2.650300e+04    11600.000000   
50%       1.467100e+05       4.535700e+04    22100.000000   
75%       3.036400e+05       7.657000e+04    39300.000000   
max       3.292782e+06       1.295455e+06   509400.000000   

       total_il_high_credit_limit  
count                6.881700e+04  
mean                 5.572240e+04  
std                  5.095845e+04  
min                  1.270000e+02  
25%                  2.288000e+04  
50%                  4.200000e+04  
75%                  7.249900e+04  
max                  1.426964e+06  

[8 rows x 76 columns]

In [104]:
# Check the balance of our target values
y.value_counts()

low_risk     68470
high_risk      347
Name: loan_status, dtype: int64

In [105]:
# Split the X and y into X_train, X_test, y_train, y_test
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,y
    ,random_state=1
    , stratify=y
)

## Data Pre-Processing

Scale the training and testing data using the `StandardScaler` from `sklearn`. Remember that when scaling the data, you only scale the features data (`X_train` and `X_testing`).

In [106]:
# Create the StandardScaler instance
from sklearn.preprocessing import StandardScaler
scaler_std = StandardScaler()

In [107]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

def Categorize_Txt_Cols(df,col_name):
    """Categorizes the columns with numbers"""
    le.fit(df[col_name])
    return le.transform(df[col_name])

def Get_Columns_With_Text(df):
    """Get's alist of text columns from a DataFrame"""
    list_of_columns = list(zip(list(df.keys()), df.dtypes.to_list()))
    return [col[0] for col in list_of_columns if col[1]==np.dtype('O')]

def Convert_Text_Cols(df):
    """Determines which columns are Text"""
    list_of_text_columns = Get_Columns_With_Text(df)
    
    for col in list_of_text_columns:
        df[col] = Categorize_Txt_Cols(df,col_name=col)
    
    return df


In [108]:
# Categorize Text Columns
X_train = Convert_Text_Cols(X_train)
X_test  = Convert_Text_Cols(X_test)


In [109]:
# Fit the Standard Scaler with the training data
# When fitting scaling functions, only train on the training dataset
X_scaler_std = scaler_std.fit(X_train)


In [110]:
# Scale the training and testing data
X_train_scaled_std = X_scaler_std.transform(X_train)
X_test_scaled_std  = X_scaler_std.transform(X_test)

## Ensemble Learners

In this section, you will compare two ensemble algorithms to determine which algorithm results in the best performance. You will train a Balanced Random Forest Classifier and an Easy Ensemble classifier . For each algorithm, be sure to complete the folliowing steps:

1. Train the model using the training data. 
2. Calculate the balanced accuracy score from sklearn.metrics.
3. Display the confusion matrix from sklearn.metrics.
4. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.
5. For the Balanced Random Forest Classifier only, print the feature importance sorted in descending order (most important feature to least important) along with the feature score

Note: Use a random state of 1 for each algorithm to ensure consistency between tests

### Balanced Random Forest Classifier

In [111]:
# Resample the training data with the BalancedRandomForestClassifier
from imblearn.ensemble import BalancedRandomForestClassifier
brf = BalancedRandomForestClassifier(n_estimators=1000, random_state=1)
brf.fit(X_train, y_train)

BalancedRandomForestClassifier(n_estimators=1000, random_state=1)

In [112]:
# Calculated the balanced accuracy score
from sklearn.metrics import balanced_accuracy_score

y_pred_brf = brf.predict(X_test)
print(f"Balanced accuracy Score: {balanced_accuracy_score(y_test, y_pred_brf):%}")

Balanced accuracy Score: 77.452047%


In [113]:
confusion_matrix(y_test, y_pred_brf)
sorted(set(y_pred_brf))

['high_risk', 'low_risk']

In [114]:
def Combine_Names_for_CF_Matrix(names=None, y_pred=None):
    if names is not None:
        lst = names
    elif y_pred is not None:
        lst = sorted(set(y_pred))

    idx = [f"Actual {x}" for x in lst]
    col = [f"Prediction {x}" for x in lst]

    return idx,col

def Generate_Confusion_Matrix(y_test, y_pred, names=None):
    from sklearn.metrics import confusion_matrix

    idx,col = Combine_Names_for_CF_Matrix(names=names, y_pred=y_pred)

    cm_df = pd.DataFrame(
        data=confusion_matrix(y_test, y_pred)
        , index= idx
        , columns= col
    )
    cm_df['Total Predictions'] = cm_df.sum(axis='columns')
    cm_df.loc["Total Actuals"] = cm_df.sum(axis='index')

    return cm_df
def Calc_Conf_Matrix(cm_df:pd.DataFrame):
    true_negative  = cm_df.iloc[0,0]
    false_positive = cm_df.iloc[0,1]
    false_negative = cm_df.iloc[1,0]
    true_positive  = cm_df.iloc[1,1]
    total_predictions  = cm_df.loc['Total Actuals','Total Predictions']
    total_predict_yes  = cm_df.iloc[2,1]
    total_predict_no     = cm_df.iloc[2,0]
    
    total_actuals_positive = cm_df.iloc[1,2]
    total_actuals_negaitive = cm_df.iloc[0,2]

    accuracy                =  (true_positive  + true_negative ) /total_predictions
    misclassification_rate  =  (false_positive + false_negative) /total_predictions  # Error Rate
    sensitivity_rate        =  true_positive / total_actuals_positive                # True Positive Rate or Recall
    false_positive_rate     =  false_positive / total_actuals_negaitive
    true_negative_rate      =  true_negative  / total_actuals_negaitive              # Specificity
    precision               =  true_positive /  total_actuals_positive

    return {'accuracy':accuracy
            ,'misclassification_rate':misclassification_rate
            ,'sensitivity_rate':sensitivity_rate
            ,'false_positive_rate':false_positive_rate
            ,'true_negative_rate':true_negative_rate
            ,'precision':precision}

def Describe_Confusion_Matrix(cm_df:pd.DataFrame):
    results = Calc_Conf_Matrix(cm_df=cm_df)
    display(cm_df)

    print(f"Overall, how often is the classifier correct (Accuracy): {results['accuracy']:.3%}")
    print(f"Overall, how often is it wrong (Misclassification Rate or Error Rate): {results['misclassification_rate']:.3%}")
    print(f"When it's actually yes, how often does it predict yes? (True Positive Rate or Sensitivity or Recall): {results['sensitivity_rate']:.3%}")
    print(f"When it's actually no, how often does it predict no (False Posive Rate): {results['true_negative_rate']:.3%}")
    print(f"When it predicts yes, how often is it correct? (Precision):{results['precision']:.3%}")
    print(f"When it's actually no, how often does it predict no? (False Positive Rate): {results['false_positive_rate']:.3%}")
    

In [115]:
# Display the confusion matrix
n=["High Risk Loans", "Low Risk Loans"]
cm_df = Generate_Confusion_Matrix(y_test=y_test, y_pred=y_pred_brf,names=n)
Describe_Confusion_Matrix(cm_df)

Prediction High Risk Loans  Prediction Low Risk Loans  \
Actual High Risk Loans                          56                         31   
Actual Low Risk Loans                         1620                      15498   
Total Actuals                                 1676                      15529   

                        Total Predictions  
Actual High Risk Loans                 87  
Actual Low Risk Loans               17118  
Total Actuals                       17205

Overall, how often is the classifier correct (Accuracy): 90.404%
Overall, how often is it wrong (Misclassification Rate or Error Rate): 9.596%
When it's actually yes, how often does it predict yes? (True Positive Rate or Sensitivity or Recall): 90.536%
When it's actually no, how often does it predict no (False Posive Rate): 64.368%
When it predicts yes, how often is it correct? (Precision):90.536%
When it's actually no, how often does it predict no? (False Positive Rate): 35.632%


In [116]:
# Print the imbalanced classification report
from imblearn.metrics import classification_report_imbalanced

print(classification_report_imbalanced(y_test,y_pred_brf, digits=5))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk    0.03341   0.64368   0.90536   0.06353   0.76339   0.56751        87
   low_risk    0.99800   0.90536   0.64368   0.94943   0.76339   0.59801     17118

avg / total    0.99313   0.90404   0.64500   0.94495   0.76339   0.59786     17205



In [117]:
# List the features sorted in descending order by feature importance
importances = brf.feature_importances_
ordered_feature_list = sorted(zip(X_train.columns,importances), key=lambda tup: tup[1], reverse=True)
_ = [print(f"Feature:\t{f[0]:30} with a value of {f[1]:10.4f}") for f in ordered_feature_list if f[1]>0.0]

Feature:	total_rec_prncp                with a value of     0.0719
Feature:	total_rec_int                  with a value of     0.0640
Feature:	total_pymnt_inv                with a value of     0.0604
Feature:	last_pymnt_amnt                with a value of     0.0603
Feature:	total_pymnt                    with a value of     0.0557
Feature:	int_rate                       with a value of     0.0294
Feature:	issue_d                        with a value of     0.0228
Feature:	dti                            with a value of     0.0185
Feature:	installment                    with a value of     0.0175
Feature:	annual_inc                     with a value of     0.0172
Feature:	max_bal_bc                     with a value of     0.0170
Feature:	mths_since_recent_inq          with a value of     0.0164
Feature:	out_prncp_inv                  with a value of     0.0161
Feature:	il_util                        with a value of     0.0160
Feature:	out_prncp                      with a value of     0.

### Easy Ensemble Classifier

In [118]:
# Train the Classifier
from imblearn.ensemble import EasyEnsembleClassifier
easy_ensem = EasyEnsembleClassifier(n_estimators=1000, random_state=1)
easy_ensem.fit(X_train,y_train)

EasyEnsembleClassifier(n_estimators=1000, random_state=1)

In [119]:
# Calculated the balanced accuracy score
y_pred_ensem = easy_ensem.predict(X_test)
print(f"Balance Accuracy Score: {balanced_accuracy_score(y_test, y_pred_ensem):%}")

Balance Accuracy Score: 92.574866%


In [120]:
# Display the confusion matrix

n=["High Risk Loans", "Low Risk Loans"]
cm_df_ensem = Generate_Confusion_Matrix(y_test=y_test, y_pred=y_pred_ensem,names=n)
Describe_Confusion_Matrix(cm_df_ensem)

Prediction High Risk Loans  Prediction Low Risk Loans  \
Actual High Risk Loans                          79                          8   
Actual Low Risk Loans                          968                      16150   
Total Actuals                                 1047                      16158   

                        Total Predictions  
Actual High Risk Loans                 87  
Actual Low Risk Loans               17118  
Total Actuals                       17205

Overall, how often is the classifier correct (Accuracy): 94.327%
Overall, how often is it wrong (Misclassification Rate or Error Rate): 5.673%
When it's actually yes, how often does it predict yes? (True Positive Rate or Sensitivity or Recall): 94.345%
When it's actually no, how often does it predict no (False Posive Rate): 90.805%
When it predicts yes, how often is it correct? (Precision):94.345%
When it's actually no, how often does it predict no? (False Positive Rate): 9.195%


In [121]:
# Print the imbalanced classification report

from imblearn.metrics import classification_report_imbalanced

print(classification_report_imbalanced(y_test, y_pred_ensem))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.08      0.91      0.94      0.14      0.93      0.85        87
   low_risk       1.00      0.94      0.91      0.97      0.93      0.86     17118

avg / total       0.99      0.94      0.91      0.97      0.93      0.86     17205



### Final Questions

1. Which model had the best balanced accuracy score?

    The ensemble method gave the best balanced accuracy score with a value of 92.57% vs. 77.45%

2. Which model had the best recall score?

    The balance random forest had a recall score of 90.536% versus 94.345% for the ensemble method.  Therefore the ensemble method has a better recall score.

3. Which model had the best geometric mean score?

    The Binary Randome Forest scored 76.34% while the Ensemble method scored 93%

4. What are the top three features?

Feature:	total_rec_prncp                with a value of     0.0719

Feature:	total_rec_int                  with a value of     0.0640

Feature:	total_pymnt_inv                with a value of     0.0604
    

In [ ]:
from collections import Counter
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix
from imblearn.ensemble import EasyEnsembleClassifier 

X, y = make_classification(
    n_classes=2, class_sep=2,
    weights=[0.1, 0.9], 
    n_informative=3, 
    n_redundant=1, 
    flip_y=0,
    n_features=20, 
    n_clusters_per_class=1, 
    n_samples=1000, 
    random_state=10
)
X,y

print('Original dataset shape %s' % Counter(y))

X_train, X_test, y_train, y_test = 
        train_test_split(X, y,
                        random_state=0)

eec = EasyEnsembleClassifier(random_state=42)
eec.fit(X_train, y_train) 

y_pred = eec.predict(X_test)
print(confusion_matrix(y_test, y_pred))